test/Controller/SubmitControllerTest.php

```diff
@@ -92,6 +92,66 @@ public function it_redirects_you_to_xpub_with_a_jwt()
         $this->assertFalse($jwt['new-session']);
     }
 
+    /**
+     * @test
+     */
+    public function it_redirects_you_to_a_trusted_url_with_a_jwt()
+    {
+        $client = static::createClient();
+        $this->logIn($client);
+
+        $client->request('GET', '/submit?return_url='.urlencode('http://localhost/path?query=arg'));
+        $response = $client->getResponse();
+
+        $this->assertTrue($response->isRedirect());
+        $location = new Uri($response->headers->get('Location'));
+
+        $this->assertSameUri('http://localhost/path?query=arg', $location->withFragment(''));
+
+        $jwt = (array) JWT::decode($location->getFragment(), $this->getParameter('xpub_client_secret'), ['HS256']);
+
+        $this->assertFalse($jwt['new-session']);
+    }
+
+    /**
+     * @test
+     */
+    public function it_redirects_you_to_a_trusted_url_with_a_jwt_in_query_argument()
+    {
+        $client = static::createClient();
+        $this->logIn($client);
+
+        $client->request('GET', '/submit?return_url='.urlencode('http://localhost/path?query=arg').'&token_in_query=true');
+        $response = $client->getResponse();
+
+        $this->assertTrue($response->isRedirect());
+        $location = new Uri($response->headers->get('Location'));
+
+        $this->assertUriStartsWith('http://localhost/path?query=arg', $location->withFragment(''));
+        $query = parse_query($location->getQuery());
+
+        $jwt = (array) JWT::decode($query['token'] ?? '', $this->getParameter('xpub_client_secret'), ['HS256']);
+
+        $this->assertFalse($jwt['new-session']);
+    }
+
+    /**
+     * @test
+     */
+    public function it_does_not_redirect_if_return_url_is_not_trusted()
+    {
+        $client = static::createClient();
+        $this->logIn($client);
+
+        $client->request('GET', '/submit?return_url='.urlencode('http://localhostevil/path?query=arg').'&token_in_query=true');
+
+        $this->assertSame(404, $client->getResponse()->getStatusCode());
+
+        $client->request('GET', '/submit?return_url='.urlencode('http://evillocalhost/path?query=arg').'&token_in_query=true');
+
+        $this->assertSame(404, $client->getResponse()->getStatusCode());
```

Comment:
The assertions are separate so should use a data provider instead (so the first one failing doesn't stop the second one from running).